File: src/components/resource-state.ts
export type ResourceState<T = Record<string, unknown>> = {
  status: "idle" | "loading" | "ready" | "error" | "unavailable";
  data?: T;
  error?: string;
  stale?: boolean;
};

export function beginResourceRefresh<T>(
  current: ResourceState<T>,
): ResourceState<T> {
  return {
    status: "loading",
    data: current.data,
  };
}

export function settleResourceRefresh<T>(
  current: ResourceState<T>,
  next: ResourceState<T>,
): ResourceState<T> {
  if (next.status !== "error" || next.data || !current.data) {
    return next;
  }
  return {
    ...next,
    data: current.data,
    stale: true,
  };
}

File: apps/flutter/lib/core/state/resource_state.dart
sealed class ResourceState<T> {
  const ResourceState();

  R when<R>({
    required R Function() idle,
    required R Function() loading,
    required R Function(T data) ready,
    required R Function(Object error) failed,
  }) => switch (this) {
    ResourceIdle<T>() => idle(),
    ResourceLoading<T>() => loading(),
    ResourceReady<T>(:final data) => ready(data),
    ResourceFailed<T>(:final error) => failed(error),
  };
}

final class ResourceIdle<T> extends ResourceState<T> {
  const ResourceIdle();
}

final class ResourceLoading<T> extends ResourceState<T> {
  const ResourceLoading();
}

final class ResourceReady<T> extends ResourceState<T> {
  const ResourceReady(this.data);
  final T data;
}

final class ResourceFailed<T> extends ResourceState<T> {
  const ResourceFailed(this.error);
  final Object error;
}
